File: src/main/java/org/trostheide/generativeArt/core/ParameterDefinition.java
package org.trostheide.generativeArt.core;

public record ParameterDefinition(
        String name,
        ParameterType type,
        Object defaultValue,
        Object min,
        Object max,
        java.util.List<String> options,
        String description) {
    // Convenience constructors
    public static ParameterDefinition integer(String name, int defaultValue, int min, int max, String description) {
        return new ParameterDefinition(name, ParameterType.INTEGER, defaultValue, min, max, null, description);
    }

    public static ParameterDefinition doubleVal(String name, double defaultValue, double min, double max,
            String description) {
        return new ParameterDefinition(name, ParameterType.DOUBLE, defaultValue, min, max, null, description);
    }

    public static ParameterDefinition bool(String name, boolean defaultValue, String description) {
        return new ParameterDefinition(name, ParameterType.BOOLEAN, defaultValue, null, null, null, description);
    }

    public static ParameterDefinition string(String name, String defaultValue, String description) {
        return new ParameterDefinition(name, ParameterType.STRING, defaultValue, null, null, null, description);
    }

    public static ParameterDefinition selection(String name, String defaultValue, java.util.List<String> options,
            String description) {
        return new ParameterDefinition(name, ParameterType.STRING, defaultValue, null, null, options, description);
    }
}

File: web/src/core/ParameterDefinition.js
export class ParameterDefinition {
    constructor(name, type, defaultValue, min, max, description, options) {
        this.name = name; // Identifier
        this.type = type; // 'integer', 'double', 'boolean', 'color', 'selection'
        this.defaultValue = defaultValue;
        this.min = min;
        this.max = max;
        this.description = description;
        this.options = options || [];
    }

    static integer(name, defaultValue, min, max, description) {
        return new ParameterDefinition(name, 'integer', defaultValue, min, max, description);
    }

    static doubleVal(name, defaultValue, min, max, description) {
        return new ParameterDefinition(name, 'double', defaultValue, min, max, description);
    }

    static string(name, defaultValue, description) {
        return new ParameterDefinition(name, 'string', defaultValue, null, null, description);
    }

    static bool(name, defaultValue, description) {
        return new ParameterDefinition(name, 'boolean', defaultValue, null, null, description);
    }

    static selection(name, defaultValue, options, description) {
        return new ParameterDefinition(name, 'selection', defaultValue, null, null, description, options);
    }

    // Add others like boolean, color later if needed
}
